src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+
+    private void run() {
+        this.ui.showGreet();
+        running:
+        while (true) {
+            try {
+                String userInput = this.ui.readInput();
+                switch (Parser.parseCmd(userInput)) {
+                case BYE:
+                    this.exit();
+                    break running;
+                case LIST:
+                    this.ui.showList(this.taskList);
+                    break;
+                case TODO:
+                    this.addTodo(Parser.parseArgs(userInput));
+                    break;
+                case EVENT:
+                    this.addEvent(Parser.parseArgs(userInput));
+                    break;
+                case DEADLINE:
+                    this.addDeadline(Parser.parseArgs(userInput));
+                    break;
+                case DONE:
+                    this.markAsDone(Parser.parseArgs(userInput));
+                    break;
+                case DELETE:
+                    this.delete(Parser.parseArgs(userInput));
+                    break;
+                case FIND:
+                    this.ui.showSearchResult(this.taskList, Parser.parseArgs(userInput));
+                    break;
+                }
+            } catch (DukeException e) {
+                this.ui.showDukeException(e);
+            }
+        }
+        this.ui.showFarewell();
+    }
+
+
+    private void exit() {
+        try {
+            this.ui.closeInput();
+            this.storage.saveData(this.taskList);
+        } catch (IOException e) {
+            this.ui.showSavingError();
+        }
+    }
+
+
+    private void addTodo(String args) throws DukeMissingArgumentException {
+        try {
+            Task todo = new Todo(args);
+            this.taskList.add(todo);
+            this.ui.showAdd(todo, this.taskList.getLength());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeMissingArgumentException();
+        }
+    }
+
+
+    private void addEvent(String args) throws DukeMissingArgumentException {
+        try {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yy HH:mm");
+            String[] splits = args.split(" /from ", 2);
+            String[] timestamps = splits[1].split(" /to ", 2);
+            String start = timestamps[0];
+            String end = timestamps[1];
+            LocalDateTime startTime = LocalDateTime.parse(start, formatter);
+            LocalDateTime endTime = LocalDateTime.parse(end, formatter);
+            if (startTime.isAfter(endTime)) {
+                this.ui.showInvalidDateRange();
+                return;
+            }
+
+            Task event = new Event(splits[0], startTime, endTime);
+            this.taskList.add(event);
+            this.ui.showAdd(event, this.taskList.getLength());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeMissingArgumentException();
+        } catch (DateTimeParseException e) {
```

Comment:
Since it seems that you have created a subtype of DukeException for each possible common errors, perhaps we could keep this consistent by creating one for this too? Since it is used twice in this file.